sdks/python/apache_beam/io/gcp/gcsfilesystem.py

```diff
@@ -0,0 +1,205 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""GCS File system for file based sources and sinks."""
+
+from __future__ import absolute_import
+
+from apache_beam.io.filesystem import CompressedFile
+from apache_beam.io.filesystem import CompressionTypes
+from apache_beam.io.filesystem import FileSystem
+from apache_beam.io.filesystem import FileMetadata
+from apache_beam.io.gcp import gcsio
+
+
+class GCSFileSystem(FileSystem):
+  """A GCS file system implementation for accessing files on disk.
+  """
+
+  def mkdirs(self, path):
+    """Recursively create directories for the provided path.
+
+    Args:
+      path: string path of the directory structure that should be created
+
+    Raises:
+      IOError if leaf directory already exists.
+    """
+    pass
```

Comment:
There isn't a concept on empty directory in GCS so you can directly write key a/b/c without pre creating a/b so this shouldn't do anything.